crates/runtime/src/memory.rs

```diff
@@ -253,9 +263,34 @@ impl Memory {
         Ok(Memory::Static {
             base,
             size: minimum,
-            make_accessible,
+            make_accessible: Some(make_accessible),
             #[cfg(all(feature = "uffd", target_os = "linux"))]
             guard_page_faults: Vec::new(),
+            #[cfg(feature = "memfd-allocator")]
+            memfd_slot: None,
+        })
+    }
+
+    /// Create a new static (immovable) memory instance backed by a MemFD.
+    #[cfg(feature = "memfd-allocator")]
+    pub fn new_memfd(
```

Comment:
Could this be merged with `new_static` and it grows various arguments to match the contents of `Memory::Static`?

IIRC the `let base = match maximum { ... }` was non-trivial and I forgot it the first time I wrote something awhile back, so having it not-duplicated I think would be best.

If there's issues with `MemFdSlot` being conditionally defined I think it would be reasonable to have something like:

```rust
#[cfg(not(feature = "memfd-allocator"))]
enum MemFdSlot {}
```

which should make `Option<MemFdSlot>` a 0-sized struct and statically assert we don't try to actually create one.

Reply:
I actually think that having this conditionally-defined type would be useful for other things as well, it would remove all the `#[cfg]` below with something like:

```rust
impl MemFdSlot {
    fn has_image(&self) -> bool { match self {} }
    fn set_heap_limit(&mut self) -> Result<()> { match self {} }
}
```

(etc)

Reply:
Ah, this refactor is so much nicer. I did this for `MemFdSlot` and also the `ModuleMemFds` below (just a zero-sized type if memfd support is not included).